plugins/cinema4d/import_cast.pyp

```diff
@@ -0,0 +1,431 @@
+
+        vertexPositions = mesh.VertexPositionBuffer()
+        vertexCount = int(len(vertexPositions) / 3)
+        
+        faces = mesh.FaceBuffer()
+        faceIndicesCount = len(faces)
+        facesCount = int(faceIndicesCount / 3)
+        faces = unpack_list([(faces[x + 2], faces[x + 1], faces[x + 0])
+                        for x in range(0, faceIndicesCount, 3)])
+        
+        newMesh.ResizeObject(vertexCount, facesCount)
+
+        # Vertice
+        for i in range(0, len(vertexPositions), 3):
+            vertexIndex = i // 3
+            x, y, z = vertexPositions[i:i+3]
+            newMesh.SetPoint(vertexIndex, Vector(x, y, -z)) # Matching the Maya default import
+
+        # Faces
+        for i in range(0, len(faces), 3):
+            polyIndex = i // 3
+            a, b, c = faces[i:i+3]
+            newMesh.SetPolygon(polyIndex, c4d.CPolygon(a, b, c))
+
+        # UVW
+        for i in range(mesh.UVLayerCount()):
+            uvTag = c4d.UVWTag(facesCount)
+            uvBuffer = mesh.VertexUVLayerBuffer(0)
+            uvUnpacked = unpack_list([(uvBuffer[x * 2], 1.0 - uvBuffer[(x * 2) + 1]) for x in faces])
+            for i in range(0, len(uvUnpacked), 6):
+                polyIndex = i // 6
+                u1, v1, u2, v2, u3, v3 = uvUnpacked[i:i+6]
+                uvTag.SetSlow(polyIndex, Vector(u1, 1-v1,0),
+                    Vector(u2, 1-v2,0),
+                    Vector(u3, 1-v3,0),
+                    Vector(0,0,0))
+            newMesh.InsertTag(uvTag)
+
+        # Vertex Colors
+        vertexColors = mesh.VertexColorBuffer()
+        if vertexColors is not None:
+            vcTag = c4d.VertexColorTag(vertexCount)
+            vertexColorList = [x for xs in [CastColor.fromInteger(x) for x in vertexColors] for x in xs], len(vertexColors) * 4
+            vcData = vcTag.GetDataAddressW()
+            for i in range(0, len(vertexColorList[0]), 4):
+                vIndex = i // 4
+                r, g, b, a = vertexColorList[0][i:i+4]
+                vcTag.SetPoint(vcData,None,None,vIndex,c4d.Vector4d(r, g, b, a))
+            newMesh.InsertTag(vcTag)
+
+        # Vertex Normals
+        vertexNormals = mesh.VertexNormalBuffer()
+        if vertexNormals is not None:
+            # Yeah I don't know what I'm doing here
+            normaltag = c4d.NormalTag(count=newMesh.GetPolygonCount())
+
+            normalListUnpacked = unpack_list([(vertexNormals[x * 3], vertexNormals[(x * 3) + 1], vertexNormals[(x * 3) + 2]) for x in faces])
+            normalList = [Vector(normalListUnpacked[i], normalListUnpacked[i+1], -normalListUnpacked[i+2]) for i in range(0, len(normalListUnpacked), 3)]
+
+            # Even if it's a Tri, you should pass a value.
+            for i in range(len(normalList) // 3):
+                normalList.insert((i + 1) * 4 - 1, Vector(0, 0, 0))
+
+            # Maps data from float to int16 value
+            normalListToSet = [int(component * 32000.0)
```

Comment:
This doesn't seem like the right logic.

Reply:
Can you explain this further? Got the idea from an older official python example:

https://github.com/PluginCafe/cinema4d_py_sdk_extended/blob/16990bb52bcfbf2cea5b66eacb685c64845859a2/scripts/04_3d_concepts/modeling/read_write_normal_tag_r13.py#L10C7-L12C34

https://github.com/PluginCafe/cinema4d_py_sdk_extended/blob/16990bb52bcfbf2cea5b66eacb685c64845859a2/scripts/04_3d_concepts/modeling/read_write_normal_tag_r13.py#L93

Reply:
Nice! I would include the snippet in the first link as a comment to explain this, because the 32000 is extremely arbitrary! It explains it better than 'float to int16'.